osu.Game.Rulesets.Osu.Tests/SpinFramesGenerator.cs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using osu.Game.Rulesets.Osu.Replays;
+using osu.Game.Rulesets.Replays;
+using osuTK;
+
+namespace osu.Game.Rulesets.Osu.Tests
+{
+    public class SpinFramesGenerator
+    {
+        /// <summary>
+        /// A small amount to spin beyond a given angle to mitigate floating-point precision errors.
+        /// </summary>
+        private const float spin_error = MathF.PI / 8;
+
+        private readonly SpinFramesGenerator? last;
+        private readonly float startAngle;
+        private readonly float endAngle;
+        private readonly double duration;
+
+        private SpinFramesGenerator(float startAngle)
+            : this(null, startAngle, 0)
+        {
+        }
+
+        private SpinFramesGenerator(SpinFramesGenerator? last, float endAngle, double duration)
+        {
+            this.last = last;
+            startAngle = last?.endAngle ?? endAngle;
+            this.endAngle = endAngle;
+            this.duration = duration;
+        }
+
+        public List<ReplayFrame> Build(double spinnerStartTime, float spinOffset)
```

Comment:
I'm also not immediately sure what `spinOffset` is and how it differs from the `.From()` angle specification. I don't think this is important if the above proposed structural change is applied though.

Reply:
It's positional offset from the centre of the playfield. So with a non-zero `spinOffset`, you aren't spinning around the centre of the playfield exactly, but around some other slightly-perturbed location relative to the centre.

Maybe rename to `spinCentreOffset` or something?

Reply:
Removed.